Assets/MixedRealityToolkit/Services/BaseDataProviderAccessCoreSystem.cs

```diff
@@ -45,21 +78,73 @@ public override void Update()
         {
             base.Update();
 
-            foreach (var provider in dataProviders)
+            // Applications may choose to trigger service changes (i.e. setting MixedRealityToolkit.ActiveProfile) in
+            // response to user input, such as a button press. Since most code runs on Unity's main thread, the collection
+            // of data providers may change mid-Update, causing an InvalidOperationException to be thrown during enumeration.
+            // To avoid unneccessary error logging, we catch the exception. It is safe to abort the loop in this fashion since
+            // it is extremely unlikely that the data providers being enumerated are the same as when we started enumeration.
+            try
             {
-                provider.Update();
+                foreach (IMixedRealityDataProvider provider in dataProviders)
+                {
+                    try
+                    {
+                        provider.Update();
+                    }
+                    catch (Exception e)
+                    {
+                        Debug.LogError($"{provider.Name}'s Update method threw {e.GetType()}");
+                    }
+                }
             }
+            catch (InvalidOperationException)
+            { }
         }
 
         /// <inheritdoc />
         public override void LateUpdate()
         {
             base.LateUpdate();
 
-            foreach (var provider in dataProviders)
+            // Applications may choose to trigger service changes (i.e. setting MixedRealityToolkit.ActiveProfile) in
+            // response to user input, such as a button press. Since most code runs on Unity's main thread, the collection
+            // of data providers may change mid-LateUpdate, causing an InvalidOperationException to be thrown during enumeration.
+            // To avoid unneccessary error logging, we catch the exception. It is safe to abort the loop in this fashion since
+            // it is extremely unlikely that the data providers being enumerated are the same as when we started enumeration.
+            try
+            {
+                foreach (IMixedRealityDataProvider provider in dataProviders)
+                {
+                    try
+                    {
+                        provider.LateUpdate();
+                    }
+                    catch (Exception e)
+                    {
+                        Debug.LogError($"{provider.Name}'s LateUpdate method threw {e.GetType()}");
+                    }
+                }
+            }
+            catch (InvalidOperationException)
```

Comment:
Is it true that if you catch the exception at this level that the loop over all providers will terminate? Or, if some exception gets thrown, does the loop continue?

Reply:
Yes, It is an exception thrown by the enumerator.

Reply:
// Applications may choose to trigger service changes (i.e. setting MixedRealityToolkit.ActiveProfile) in
            // response to user input, such as a button press. Since most code runs on Unity's main thread, the collection
            // of data providers may change mid-Update, causing an InvalidOperationException to be thrown during enumeration.
            // To avoid unneccessary error logging, we catch the exception. It is safe to abort the loop in this fashion since
            // it is extremely unlikely that the data providers being enumerated are the same as when we started enumeration.